Android/android.py

```diff
@@ -74,41 +78,62 @@ def subdir(*parts, create=False):
 
 def run(command, *, host=None, env=None, log=True, **kwargs):
     kwargs.setdefault("check", True)
+
     if env is None:
         env = os.environ.copy()
-    original_env = env.copy()
-
     if host:
-        env_script = ANDROID_DIR / "android-env.sh"
-        env_output = subprocess.run(
-            f"set -eu; "
-            f"HOST={host}; "
-            f"PREFIX={subdir(host)}/prefix; "
-            f". {env_script}; "
-            f"export",
-            check=True, shell=True, text=True, stdout=subprocess.PIPE
-        ).stdout
-
-        for line in env_output.splitlines():
-            # We don't require every line to match, as there may be some other
-            # output from installing the NDK.
-            if match := re.search(
-                "^(declare -x |export )?(\\w+)=['\"]?(.*?)['\"]?$", line
-            ):
-                key, value = match[2], match[3]
-                if env.get(key) != value:
-                    print(line)
-                    env[key] = value
-
-        if env == original_env:
-            raise ValueError(f"Found no variables in {env_script.name} output:\n"
-                             + env_output)
+        env.update(android_env(host))
 
     if log:
         print(">", " ".join(map(str, command)))
     return subprocess.run(command, env=env, **kwargs)
 
 
+def print_env(context):
+    android_env(getattr(context, "host", None))
+
+
+def android_env(host):
+    if host:
+        prefix = subdir(host) / "prefix"
+    else:
+        prefix = ANDROID_DIR / "prefix"
+        sysconfigdata_files = prefix.glob("lib/python*/_sysconfigdata__android_*.py")
+        host = re.fullmatch(
+            r"_sysconfigdata__android_(.+).py", next(sysconfigdata_files).name
+        )[1]
+
+    env_script = ANDROID_DIR / "android-env.sh"
+    env_output = subprocess.run(
+        f"set -eu; "
+        f"export HOST={host}; "
+        f"PREFIX={prefix}; "
+        f". {env_script}; "
+        f"export",
+        check=True, shell=True, capture_output=True, text=True,
+    ).stdout
+
+    env = {}
+    for line in env_output.splitlines():
+        # We don't require every line to match, as there may be some other
+        # output from installing the NDK.
+        if match := re.search(
+            "^(declare -x |export )?(\\w+)=['\"]?(.*?)['\"]?$", line
+        ):
+            key, value = match[2], match[3]
+            if os.environ.get(key) != value:
+                env[key] = value
```

Comment:
This seems simpler, unless there's a need to minimise the size of 'env'? You could even use a dictcomp.
```suggestion
            key, value = match[2], match[3]
            env[key] = value
```

Reply:
This code is run by the `android.py env` command, which cibuildwheel uses to determine the environment variables it needs when building for Android. It outputs all variables changed by the android-env.sh script, which looks something like this:

```
% ./Android/android.py env aarch64-linux-android
export AR=/Users/msmith/Library/Android/sdk/ndk/27.1.12297006/toolchains/llvm/prebuilt/darwin-x86_64/bin/llvm-ar
export AS=/Users/msmith/Library/Android/sdk/ndk/27.1.12297006/toolchains/llvm/prebuilt/darwin-x86_64/bin/llvm-as
export CC=/Users/msmith/Library/Android/sdk/ndk/27.1.12297006/toolchains/llvm/prebuilt/darwin-x86_64/bin/aarch64-linux-android24-clang
export CFLAGS='-D__BIONIC_NO_PAGE_SIZE_MACRO -I/Users/msmith/git/python/cpython/cross-build/aarch64-linux-android/prefix/include'
export CPU_COUNT=10
export CXX=/Users/msmith/Library/Android/sdk/ndk/27.1.12297006/toolchains/llvm/prebuilt/darwin-x86_64/bin/aarch64-linux-android24-clang++
export CXXFLAGS='-D__BIONIC_NO_PAGE_SIZE_MACRO -I/Users/msmith/git/python/cpython/cross-build/aarch64-linux-android/prefix/include'
export HOST=aarch64-linux-android
export LD=/Users/msmith/Library/Android/sdk/ndk/27.1.12297006/toolchains/llvm/prebuilt/darwin-x86_64/bin/ld
export LDFLAGS='-Wl,--build-id=sha1 -Wl,--no-rosegment -Wl,-z,max-page-size=16384 -Wl,--no-undefined -lm -L/Users/msmith/git/python/cpython/cross-build/aarch64-linux-android/prefix/lib'
export NM=/Users/msmith/Library/Android/sdk/ndk/27.1.12297006/toolchains/llvm/prebuilt/darwin-x86_64/bin/llvm-nm
export PKG_CONFIG='pkg-config --define-prefix'
export PKG_CONFIG_LIBDIR=/Users/msmith/git/python/cpython/cross-build/aarch64-linux-android/prefix/lib/pkgconfig
export RANLIB=/Users/msmith/Library/Android/sdk/ndk/27.1.12297006/toolchains/llvm/prebuilt/darwin-x86_64/bin/llvm-ranlib
export READELF=/Users/msmith/Library/Android/sdk/ndk/27.1.12297006/toolchains/llvm/prebuilt/darwin-x86_64/bin/llvm-readelf
export SHLVL=2
export STRIP=/Users/msmith/Library/Android/sdk/ndk/27.1.12297006/toolchains/llvm/prebuilt/darwin-x86_64/bin/llvm-strip
```

If this included every single environment variable, it would be a lot less convenient to use it interactively when debugging build problems. It would also be a lot more difficult to verify its correctness.